### PR TITLE
lib: at_host: use ncs,at-host-uart

### PR DESCRIPTION
**Pull request by gmarull**
Move AT Host UART choice to Devicetree by using ncs,at-host-uart. This
allows to obtain the UART device reference at compile time.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

### CODE REVIEW

File: lib/at_host/at_host.c [hunk 1]
```diff
@@ -39,15 +35,8 @@ enum term_modes {
...
-/** @brief UARTs. */
-enum select_uart {
-	UART_0,
-	UART_1,
-	UART_2
-};
-
 static enum term_modes term_mode;
-static const struct device *uart_dev;
+static const struct device *uart_dev = DEVICE_DT_GET(DT_CHOSEN(ncs_at_host_uart));
```
Comment (anangl) [inline lib/at_host/at_host.c:39]:
```suggestion
static const struct device *const uart_dev = DEVICE_DT_GET(DT_CHOSEN(ncs_at_host_uart));
```

**Review by anangl — APPROVED** [review]

**Review by glarsennordic — CHANGES_REQUESTED** [review]
I suggest defaulting to `DT_CHOSEN(zephyr_shell_uart)` if `ncs,at-host-uart` is not set.

That way samples aren't forced to include a device tree overlay file for every board they intend to support, and this module will work on any board that supports the zephyr UART shell without need of device tree configuration changes

You can use `DT_HAS_CHOSEN` to check whether `ncs,at-host-uart` is set

**Comment by gmarull** [comment]
> I suggest defaulting to `DT_CHOSEN(zephyr_shell_uart)` if `ncs,at-host-uart` is not set.
> 
> That way samples aren't forced to include a device tree overlay file for every board they intend to support, and this module will work on any board that supports the zephyr UART shell without need of device tree configuration changes
> 
> You can use `DT_HAS_CHOSEN` to check whether `ncs,at-host-uart` is set

I don't think it is a good idea to re-use other choices. `zephyr,shell-uart` is specifically meant for the shell subsystem, so using it for something else is wrong IMO. Why should my shell be polluted with other stuff? One option would to default to, e.g., `uart0`, but I don't think this is a good practice anyway.

@mbolivar-nordic can you comment, please?

**Comment by glarsennordic** [comment]
> Why should my shell be polluted with other stuff? 

I look at `zephyr,shell-uart` as a "sane choice" for a default in situations where `ncs,at-host-uart` was not specified

Both are sharing `uart0` on nordic boards anyways, so that pollution is happening either way. And the user still has the option to set `ncs,at-host-uart` if they want both to be on separate UART devices

I personally think requiring an overlay file for every combination of sample and target board should be avoided.

I suppose you could add `ncs,at-host-uart` to the `nrf9160dk_nrf9160_common.dts` and `thingy91_nrf9160_common.dts` files, (which is where the choices for `zephyr,shell-uart`, `zephyr,console` and `zephyr,uart-mcumgr` all live). But that seems a bit too broad in scope since this library is optional for those boards

I wonder if there is a way to create default device tree overlays at the library level. You could put the choice definition there, in that case

**Comment by tokangas** [comment]
I agree with @glarsennordic that having to set `ncs,at-host-uart` in all samples it not very convenient.

I guess setting it in the board configuration for nRF9160 DK might not be possible, because the board DTS file comes from Zephyr and this is an NCS specific configuration?

**Comment by glarsennordic** [comment]
> I guess setting it in the board configuration for nRF9160 DK might not be possible, because the board DTS file comes from Zephyr and this is an NCS specific configuration?

Oh man... You're totally right. I completely forgot about that. I guess that approach is out the window.

File: doc/nrf/libraries/modem/at_host.rst [hunk 2]
```diff
@@ -24,11 +24,15 @@ The various configuration options used in the library are listed below:
 * :kconfig:option:`CONFIG_AT_HOST_CMD_MAX_LEN` - Sets the maximum length for an AT command
 * :kconfig:option:`CONFIG_AT_HOST_THREAD_PRIO` - Sets the priority level for the AT host work queue thread
 
-Configuration options for setting the UART:
+The UART is chosen in Devicetree by using ``ncs,at-host-uart`` like this:
```
Comment (divipillai) [inline doc/nrf/libraries/modem/at_host.rst:27]:
```suggestion
The UART is chosen by using ``ncs,at-host-uart`` in devicetree in the following way:
```
Just a minor suggestion.

**Comment by gmarull** [comment]
> I agree with @glarsennordic that having to set `ncs,at-host-uart` in all samples it not very convenient.
> 
> I guess setting it in the board configuration for nRF9160 DK might not be possible, because the board DTS file comes from Zephyr and this is an NCS specific configuration?

Note that this approach has been taken by other subsystems in NCS. Some action will be required for the next upmerge, as labels have been removed upstream.

**Comment by mbolivar-nordic** [comment]
> I don't think it is a good idea to re-use other choices. [...] @mbolivar-nordic can you comment, please?

I agree.

**Comment by glarsennordic** [comment]
Is it just best practice, then, to require a device tree overlay file for every combination of sample and supported board, or is there a third option we're not seeing to this problem, @mbolivar-nordic ?

**Review by divipillai — APPROVED** [review]

**Comment by trantanen** [comment]
> > One option would to default to, e.g., uart0, but I don't think this is a good practice anyway.
> 
> To be honest, I'm not sure this would actually be a terrible idea

Is this option something we should still consider? I agree with @tokangas and @glarsennordic that having some default that is used, e.g., 95% of the time would be convenient. I understand this may not be perfect ideologically but are there other drawbacks in this approach?

**Comment by plskeggs** [comment]
> I don't think it is a good idea to re-use other choices. `zephyr,shell-uart` is specifically meant for the shell subsystem, so using it for something else is wrong IMO. Why should my shell be polluted with other stuff? One option would to default to, e.g., `uart0`, but I don't think this is a good practice anyway.

FYI, I've tried using at_host and the shell together before, and it doesn't work -- they each want to handle input from the user.

I instead added a shell command to switch in and out of AT mode with `at enable`.  Enabling is done using `shell_set_root_cmd("at")`.  When in AT mode, all shell input gets routed to the nrf_modem_at_cmd() function inside my AT shell command handler.  If the user input is actually the string "exit", then it exits AT mode.

**Comment by nordicjm** [comment]
> > I guess setting it in the board configuration for nRF9160 DK might not be possible, because the board DTS file comes from Zephyr and this is an NCS specific configuration?
> 
> Oh man... You're totally right. I completely forgot about that. I guess that approach is out the window.

We have noup patches for boards in sdk-zephyr already, e.g. https://github.com/nrfconnect/sdk-zephyr/commit/c89343ac27c14bdacb61b1464c518543a0617ccd

**Comment by gmarull** [comment]
The approach listed here has been adopted by a few other NCS subsystems, not sure why this case should be an exception. I'd like to see an alternative if this is not convincing (which I can understand).

**Comment by tokangas** [comment]
I think we can say, that AT host almost always uses `uart0`. This PR has 20 board overlays, which all set the UART to `uart0`. Because of this it would be convenient that this would be the default. That's the only comment I have, otherwise this change is perfectly fine.

**Comment by de-nordic** [comment]
> I think we can say, that AT host almost always uses `uart0`. This PR has 20 board overlays, which all set the UART to `uart0`. Because of this it would be convenient that this would be the default. That's the only comment I have, otherwise this change is perfectly fine.

@gmarull Is it possible to make the uart0 default option when not provided, and document this?

**Comment by gmarull** [comment]
> > I think we can say, that AT host almost always uses `uart0`. This PR has 20 board overlays, which all set the UART to `uart0`. Because of this it would be convenient that this would be the default. That's the only comment I have, otherwise this change is perfectly fine.
> 
> @gmarull Is it possible to make the uart0 default option when not provided, and document this?

I guess it's a fair solution provided this is non-portable code. I will update PR later.

**Comment by carlescufi** [comment]
@gmarull I think the `noup` patch proposal from @nordicjm makes sense, what is it you don't like there?

**Comment by gmarull** [comment]
> @gmarull I think the `noup` patch proposal from @nordicjm makes sense, what is it you don't like there?

at_host library is using Kconfig to specify a hardware choice (hardcoding uart selection 0,1,2). This belongs to the Devicetree domain.

**Comment by carlescufi** [comment]
> > @gmarull I think the `noup` patch proposal from @nordicjm makes sense, what is it you don't like there?
> 
> at_host library is using Kconfig to specify a hardware choice (hardcoding uart selection 0,1,2). This belongs to the Devicetree domain.

I mean a `noup` that adds the `chosen` node to the upstream boards' `.dts` files @gmarull

**Comment by de-nordic** [comment]
@glarsennordic Can you take a look now? I think that your issue has been addressed.

**Comment by gmarull** [comment]
> > > @gmarull I think the `noup` patch proposal from @nordicjm makes sense, what is it you don't like there?
> > 
> > 
> > at_host library is using Kconfig to specify a hardware choice (hardcoding uart selection 0,1,2). This belongs to the Devicetree domain.
> 
> I mean a `noup` that adds the `chosen` node to the upstream boards' `.dts` files @gmarull

Not a bad solution, just that I prefer to avoid `noups`. We probably need a standard mechanism to _extend_ boards downstream.

**Comment by carlescufi** [comment]
> Not a bad solution, just that I prefer to avoid `noups`. We probably need a standard mechanism to _extend_ boards downstream.

We do. @tejlmand this is something for you mid-term.

**Review by glarsennordic — APPROVED** [review]
LGTM

Thanks for finding a way to remove all the overlays, I appreciate it

If you decide to go with a noup for the default instead, I'd be fine with that too.

**Pull request merged**